### PR TITLE
[CALCITE-5695] Add MAP_KEYS, MAP_VALUES function (enabled in Spark li…

### PR DESCRIPTION
**Pull request by liuyongvs**
MAP_KEYS - Returns an unordered array containing the keys of the map.
MAP_VALUES - Returns an unordered array containing the values of the map.
For more details
https://spark.apache.org/docs/latest/sql-ref-functions-builtin.html

### CODE REVIEW

**Comment by liuyongvs** [comment]
hi @JiajunBernoulli could you also help review it?

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 1]
```diff
@@ -3797,6 +3798,16 @@ public static List distinct(List list) {
     return new ArrayList<>(result);
   }
 
+  /** Support the MAP_KEYS function. */
+  public static List mapKeys(Map map) {
+    return (List) map.keySet().stream().collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (zoudan) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3803]: Then we should handle null here? It seems a NPE Exception will be thrown?

[the remark this answers is not in the document]
Comment (liuyongvs) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3803]: @zoudan not needs. the runtime will does it like other engine

[the remark this answers is not in the document]
Comment (liuyongvs) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3803]: and why i doesn't add unit test map_keys(cast(null as map[keyType, valueType])) is because the calcite doesn't supported, which you can see here https://github.com/apache/calcite/pull/3189
java.lang.RuntimeException: Error while parsing query: select cast(map['washington', 1, null, 44] as map[CHAR(10), INTEGER NOT NULL>] from (values (1))
	at org.apache.calcite.sql.test.AbstractSqlTester.parseAndValidate(AbstractSqlTester.java:157)
	at org.apache.calcite.sql.test.AbstractSqlTester.validateAndThen(AbstractSqlTester.java:248)
	at org.apache.calcite.sql.test.AbstractSqlTester.checkColumnType(AbstractSqlTester.java:177)
	at org.apache.calcite.test.SqlValidatorFixture.columnType(

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 2]
```diff
@@ -5361,6 +5361,28 @@ private static void checkIf(SqlOperatorFixture f) {
     f.checkNull("array_length(null)");
   }
 
+  /** Tests {@code MAP_KEYS} function from Spark. */
+  @Test void testMapKeysFunc() {
+    final SqlOperatorFixture f = fixture()
+        .setFor(SqlLibraryOperators.MAP_KEYS)
```
Comment (tanclary) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:5367]: I think it could be good to write a test that fails if library is not set to Spark (you only want it to work for Spark).

File: site/_docs/reference.md [hunk 3]
```diff
@@ -2723,6 +2723,8 @@ BigQuery's type system uses confusingly different names for types and functions:
 | m | TO_BASE64(string)                              | Converts the *string* to base-64 encoded form and returns a encoded string
 | b m | FROM_BASE64(string)                          | Returns the decoded result of a base-64 *string* as a string
 | b o | LTRIM(string)                                | Returns *string* with all blanks removed from the start
+| s | MAP_KEYS(map)                                  | Returns an unordered array containing the keys of the map.
```
Comment (julianhyde) [inline site/_docs/reference.md:2726]: remove '.'

Comment (liuyongvs) [reply site/_docs/reference.md]: fixed @julianhyde

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 2 above)
Comment (liuyongvs) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: added @tanclary

**Comment by liuyongvs** [comment]
hi @julianhyde @tanclary @zoudan  fix all your reviews, please have a look again,thanks. and i will supports all the spark collection functions latter.

**Comment by JiajunBernoulli** [comment]
> hi @JiajunBernoulli do you have time to review it ?

I usually only have time on weekends.

So I will review it tomorrow.

**Comment by julianhyde** [comment]
@liuyongvs Can you respond to my comment in Jira? What is an 'unordered array'?

File: site/_docs/reference.md [hunk 4]
```diff
@@ -2723,6 +2723,8 @@ BigQuery's type system uses confusingly different names for types and functions:
 | m | TO_BASE64(string)                              | Converts the *string* to base-64 encoded form and returns a encoded string
 | b m | FROM_BASE64(string)                          | Returns the decoded result of a base-64 *string* as a string
 | b o | LTRIM(string)                                | Returns *string* with all blanks removed from the start
+| s | MAP_KEYS(map)                                  | Returns an unordered array containing the keys of the map
+| s | MAP_VALUES(map)                                | Returns an unordered array containing the values of the map
```
Comment (JiajunBernoulli) [inline site/_docs/reference.md:2727]: map should be `*map*`.

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java [hunk 5]
```diff
@@ -536,6 +536,26 @@ public static SqlCall stripSeparator(SqlCall call) {
...
+      ARG0.andThen(SqlTypeTransforms.TO_MAP_KEYS)
+          .andThen(SqlTypeTransforms.TO_NULLABLE);
+
+  /**
+   * Returns a ARRAY type.
+   *
+   * <p>For example, given {@code (INTEGER, DATE) MAP}, returns
+   * {@code DATE ARRAY}.
+   */
+  public static final SqlReturnTypeInference TO_MAP_VALUES_NULLABLE =
```
Comment (JiajunBernoulli) [inline core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:555]: We should have two types:  *not null*  and *nullable*.
```
  public static final SqlReturnTypeInference TO_MAP_KEYS =
      ARG0.andThen(SqlTypeTransforms.TO_MAP_KEYS);

  public static final SqlReturnTypeInference TO_MAP_KEYS_NULLABLE =
      TO_MAP_KEYS.andThen(SqlTypeTransforms.TO_NULLABLE);
```

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 6]
```diff
@@ -3797,6 +3798,16 @@ public static List distinct(List list) {
...
+  }
+
+  /** Support the MAP_VALUES function. */
+  public static List mapValues(Map map) {
+    return (List) map.values().stream().collect(Collectors.toList());
```
Comment (JiajunBernoulli) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3808]: `return new ArrayList<>(map.values());` 

I think it's clearer and there won't be NPE.

**Comment by JiajunBernoulli** [comment]
I left some comments for the PR.

And we should discuss 'unordered array' in JIRA.

**Comment by liuyongvs** [comment]
fix all your review @JiajunBernoulli 
and the spark doesn't support MULTISET type.  an unordered array  means the element doesn't sort.

the spark doesn't support MULTISET type.  an unordered array  means the element doesn't sort.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3803]: Why not use `new ArrayList<>(map.keySets())`?

Comment (JiajunBernoulli) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3803]: Why not use `new ArrayList<>(map.keySet())?`

Comment (liuyongvs) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: fixed @JiajunBernoulli

**Comment by liuyongvs** [comment]
fixed , thanks for your details review @JiajunBernoulli. do you also have any comments. 
If not, i will squash the commits.

**Comment by liuyongvs** [comment]
have squashed the commits,  could anyone merge it thanks @JiajunBernoulli @julianhyde @zoudan

**Comment by JiajunBernoulli** [comment]
Conflicting files because https://github.com/apache/calcite/pull/3198

Please resolve them.

**Comment by liuyongvs** [comment]
hi @JiajunBernoulli fixed the conflict and ci passed

**Comment by JiajunBernoulli** [comment]
Please delete '.' in commit message.

**Comment by liuyongvs** [comment]
@JiajunBernoulli fixed all the current prs message ending with '.'

**Review by JiajunBernoulli — APPROVED** [review]
